### ISSUE TITLE
Help Needed: Using solid-client-authn-browser's login method for E2E Testing

### ISSUE THREAD

**Issue by bejow** [inrupt/solid-client-authn-js#2568]
Hello, I'm currently trying to test my react application that requires a solid pod for the login.

I'm stuck trying to authenticate against login.inrupt.com (without showing an UI) by using solid-client-auth-browser. My goal is to receive an access token, that I can use for testing my app automatically as a authenticated user (accessing pod data).

So what I found online was a tutorial for solid-client-authn-node:
https://docs.inrupt.com/developer-tools/javascript/client-libraries/tutorial/authenticate-nodejs-script/

That seems to do what I want, but my problem is that cypress runs in a browser environment, so i can't use this library since I get the error:
`Module not found: Error: Can't resolve 'v8' in 'C:\...\node_modules\openid-client\lib\helpers`

When I try to do the same with solid-client-authn-browser library I run into a different Problem;
The Request is sent to login.inrupt but the Response contains this error:
{"error":"invalid_grant","error_description":"Invalid redirect_uri"}

My code looks like this:
```
const { Session } = require("@inrupt/solid-client-authn-browser")
describe("Login", () => {
    const session = new Session()
    session.handleIncomingRedirect(() => {...})
    session.login({
        oidcIssuer: "https://login.inrupt.com",
        clientId: "[clientId]",
        clientSecret: "[clientSecret]",
    }).then(() => {...})
}
```

Can anyone point me into a direction or knows the solution for my problem?

**Comment by jeswr** [comment]
As documented [here](https://docs.inrupt.com/developer-tools/api/javascript/solid-client-authn-browser/classes/Session.html#login) you *must* provide a `redirectUrl` when calling `.login` in `@inrupt/solid-client-authn-browser`. Cypress should then be redirected to the `redirectUrl` once you have been authenticated and the authn library can complete the login flow from there.

**Comment by bejow** [comment]
Thank you for your help. I tried to add the redirectUrl:

```
const { Session } = require("@inrupt/solid-client-authn-browser")
describe("Login", () => {
    const session = new Session()
    session.handleIncomingRedirect(() => {...})
    session.login({
        oidcIssuer: "https://login.inrupt.com",
        clientId: "[clientId]",
        clientSecret: "[clientSecret]",
        redirectUrl: "http://localhost:3000",
    }).then(() => {...})
}
```

This still gives me the Error:
`{"error":"invalid_grant","error_description":"Invalid redirect_uri"}`

Are there constraints regarding the redirectUrl?

The Network request sent from the library looks like this:
```
fetch("https://login.inrupt.com/authorization?client_id=***&redirect_uri=http%3A%2F%2F127.0.0.1%3A3000%2F&response_type=code&scope=openid%20offline_access%20webid&state=***&code_challenge=***&code_challenge_method=S256&prompt=consent&response_mode=query", {
  "headers": {
    "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
    "accept-language": "en-US,en;q=0.9",
    "sec-ch-ua": "\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Google Chrome\";v=\"108\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"Windows\"",
    "sec-fetch-dest": "iframe",
    "sec-fetch-mode": "navigate",
    "sec-fetch-site": "cross-site",
    "upgrade-insecure-requests": "1"
  },
  "referrer": "http://127.0.0.1:3000/",
  "referrerPolicy": "strict-origin-when-cross-origin",
  "body": null,
  "method": "GET",
  "mode": "cors",
  "credentials": "omit"
});
```

**Comment by NSeydoux** [comment]
Hi @bejow, thanks for reaching out. First, let me rephrase my understanding of what you are trying to do to make sure we are on the same page: you want to run automated tests against a live Pod for a in-browser app, and in particular to handle login in this context, right?

There are different environment-specific dependencies between `@inrupt/solid-client-authn-browser` and `@inrupt/solid-client-authn-browser`, so you will not be able to use one in the other environment. If cypress runs in a browser context, you'll have to use `@inrupt/solid-client-authn-browser`.

If I understand correctly, the issue you are trying to solve is headless authentication. To do so, you are trying to use static client registration (what is documented [here](https://docs.inrupt.com/developer-tools/javascript/client-libraries/tutorial/authenticate-nodejs-script/)). However, static client registration is not supported by `@inrupt/solid-client-authn-browser`, as it is only suited to a backend context. Without getting into too much details, a client ID/client secret pair identifies a single instance of your program, which works well for a server, but doesn't work for browser-side code, which will have as many instances as there are users running it (and also, the browser context is not secure to send the ID and secret).

The problem you are trying to solve is something we face in our own browser-based tests. The way we approached it was to set environment variables for the user login and password (not client ID and secret), and have the test driver go through the user flow of clicking the "login" button, go to the OpenID provider, enter the login and password, and get redirected back to the app being tested to complete authentication. We are using playwright instead of cypress, so your code will end up looking different, but you can have a peek at https://github.com/inrupt/typescript-sdk-tools/blob/main/packages/internal-playwright-helpers/src/flows/auth.ts to see how we do this.

Would this approach work for you?

**Comment by bejow** [comment]
Hey @NSeydoux your answer is much appreciated. And yes you got my Question right. 

This approach will work for me and I already started to go down that way, since I wasn't able to get it done the headless way. 

Do you think there will be headless authentication option for the @inrupt/solid-client-authn-browser in the future?
Would it make sense/work to have something like this: 
`const {token} = login(oidcIssuer ,username, password)` ?

Anyway thank you very much @NSeydoux & @jeswr for your work & fast replies - I think this Issue can be closed then!

**Comment by NSeydoux** [comment]
There is no standard and secure OIDC flow to do what you describe: one of the core OpenID principles is that the app doesn't get the user credentials, which is essentially what headless authentication would look like. For this reason, this will likely never be supported by the library.

I'm glad you could make it work another way!
